worlds/lufia2ac/Client.py

```diff
@@ -0,0 +1,360 @@
+    0xCE: "a Goblin",
+    0xCF: "a Goblin Mage",
+    0xD0: "a Slave",
+    0xD1: "a Follower",
+    0xD2: "a Groupie",
+    0xD3: "the Egg Dragon",
+    0xD4: "a Mummy",
+    0xD5: "a Troll",
+    0xD6: "Gades",
+    0xD7: "Idura",
+    0xD8: "a Lion",
+    0xD9: "the Rogue Flower",
+    0xDA: "a Gargoyle",
+    0xDB: "a Ghost Ship",
+    0xDC: "Idura",
+    0xDD: "a Soldier",
+    0xDE: "Gades",
+    0xDF: "the Master",
+}
+
+
+class L2ACSNIClient(SNIClient):
+    game: str = "Lufia II Ancient Cave"
+
+    async def validate_rom(self, ctx: SNIContext) -> bool:
+        from SNIClient import snes_read
+
+        rom_name: bytes = await snes_read(ctx, L2AC_ROMNAME_START, 0x15)
+        if rom_name is None or rom_name[:4] != b"L2AC":
+            return False
+
+        ctx.game = self.game
+        ctx.items_handling = 0b111  # fully remote
+
+        ctx.rom = rom_name
+
+        return True
+
+    async def game_watcher(self, ctx: SNIContext) -> None:
+        from SNIClient import snes_buffered_write, snes_flush_writes, snes_read
+
+        rom: bytes = await snes_read(ctx, L2AC_ROMNAME_START, 0x15)
+        if rom != ctx.rom:
+            ctx.rom = None
+            return
+
+        if ctx.server is None or ctx.slot is None:
+            # not successfully connected to a multiworld server, cannot process the game sending items
+            return
+
+        signature: bytes = await snes_read(ctx, L2AC_SIGN_ADDR, 16)
+        if signature != b"ArchipelagoLufia":
+            return
+
+        # Goal
+        if not ctx.finished_game:
+            goal_data: bytes = await snes_read(ctx, L2AC_GOAL_ADDR, 10)
+            if goal_data is not None and goal_data[goal_data[0]] == 0x01:
+                await ctx.send_msgs([{"cmd": "StatusUpdate", "status": ClientStatus.CLIENT_GOAL}])
+                ctx.finished_game = True
+
+        # DeathLink TX
+        death_data: bytes = await snes_read(ctx, L2AC_DEATH_ADDR, 3)
+        if death_data is not None:
+            await ctx.update_death_link(bool(death_data[0]))
+            if death_data[1] != 0x00:
+                snes_buffered_write(ctx, L2AC_DEATH_ADDR + 1, b"\x00")
+                if "DeathLink" in ctx.tags and ctx.last_death_link + 1 < time.time():
+                    player_name: str = ctx.player_names.get(ctx.slot, str(ctx.slot))
+                    enemy_name: str = enemy_names.get(death_data[1] - 1, hex(death_data[1] - 1))
+                    await ctx.send_death(f"{player_name} was totally defeated by {enemy_name}.")
+
+        # TX
+        tx_data: bytes = await snes_read(ctx, L2AC_TX_ADDR, 8)
+        if tx_data is not None:
+            snes_items_sent = int.from_bytes(tx_data[:2], "little")
+            client_items_sent = int.from_bytes(tx_data[2:4], "little")
+            client_ap_items_found = int.from_bytes(tx_data[4:6], "little")
+
+            if client_items_sent < snes_items_sent:
+                location_id: int = locations_start_id + client_items_sent
```

Comment:
The Chest number is based on the order you opened them in. Every time you open a chest, it is the next one in the sequence. When you die and start over, it continues counting from the last chest you opened.

2 players would not get the same dungeons, as the dungeons are randomized by the game.

I believe keeping track of your chest number is tied to your save file (the SNES SRAM), so 2 players connecting to the same world would be overlapping their chests, rather than new ones. Co-op does not seem designed for this game.

Reply:
> > If two people connect to the same slot, with personA opening four chests and personB opening a single one, the checked locations on the server should then be "Blue chest 1" to "Blue chest 4", but not "Blue chest 5".
> 
> Out of curiosity, is this an intentional design decision? I could see some players being interested in being able to co-op a seed (two players playing the same slot). Especially since, as I understand it, some of Lufia's goals/options can be quite long. Being able to co-op a slot would make that more "sync-viable".

There is only one mode that can be shortened by co-op-ing the game and that would be Iris Treasure Hunt. Iris Treasures are red chest items, and therefore the blue chest count or content is irrelevant for this mode. The issue I see with co-op-ing that mode is that I am not sure how Iris Treasures are handled. @el-u will have more insight here...